Implement RelatIF for Explicit, CG, and LISSA

jxbb824 opened the pull request:
## Description  

### 1. Motivation and Context  
This PR implements two versions of RelatIF as described in [https://arxiv.org/pdf/2003.11630](https://arxiv.org/pdf/2003.11630). Instead of creating new attributors, the implementation integrates RelatIF into the existing influence function attributors via a new config parameter.  

This addresses issue **#164**.

### 2. Summary of the change  
- Implemented two versions of RelatIF for the **Explicit**, **CG**, and **LISSA** influence function attributors.  
- Introduced a new config parameter to enable RelatIF computation within the existing framework.  
- Added an example that replicates the first figure from the RelatIF paper to demonstrate the correctness of the implementation.  

### 3. What tests have been added/updated for the change?  
- [ ] N/A: No test will be added (please justify)  
- [x] Unit test: Added tests to verify RelatIF computation in explicit, CG, and LISSA attributors.  
- [x] Application test: Included an example script that reproduces the first figure from the RelatIF paper.  
- [ ] Document test: Verified that the new method/config parameter is correctly documented.

TheaperDeng commented on dattri/algorithm/influence_function.py:
```diff
@@ -289,6 +289,49 @@ def transform_test_rep(
+        Raises:
+            ValueError: If an unknown `method` is provided.
+        """
+        transformed = self.transform_test_rep(ckpt_idx, train_batch_rep)
```
I don't know if this `transformed` can be used for future computation just to save the computational resource. (I mean it definitely can, but we may have a messy code base after that) **But I think it's fine to put it as it is right now**.

jxbb824 replied:
Thanks for the feedback! I'll keep it as is for now.

TheaperDeng commented on dattri/algorithm/base.py:
```diff
@@ -314,6 +321,15 @@ def attribute(
+                if relativeif_method is not None:
+                    denom = self._compute_denom(
+                        checkpoint_idx,
+                        train_batch_rep,
```
I think you need 2 input here. One is this `train_batch_rep` comes from `train_batch_rep = self.generate_train_rep(..., train_batch_data, ...)` and another `test_batch_rep` comes from `self.generate_test_rep(..., train_batch_data, ...)`. The first one will be used for H^-1g_i and the second one will be used as g_i when `relativeif_method == "l"`. Since sometimes there is a difference between loss function and target function.

jxbb824 replied:
Thanks for the explanation! Since the implementation in the Relatif paper doesn’t use `s_test`, it seems that the roles of `transform_test_rep` and `transform_train_rep` are effectively reversed compared to how other influence functions are implemented in the library. While this doesn’t cause any issues for now, it could introduce potential problems if we later extend this to other methods where `transform_train_rep` has an actual purpose. It might also lead to some confusion in understanding the code.

Additionally, the naming of `self.generate_test_rep(..., train_batch_data, ...)` can be a bit misleading. For now, I’m just calling it `test_batch_rep`.

TheaperDeng reviewed (commented):
good job. Please change according to the comments. Another problem is that how much overhead does it involve when open the `relatif_method` config to `l` and `theta`.

TheaperDeng commented:
Oh another minor comment is that you may add RelatIF here: https://github.com/TheaperDeng/dattri?tab=readme-ov-file#algorithms-supported and remove the TODO here: https://github.com/TRAIS-Lab/dattri?tab=readme-ov-file#development-plan

TheaperDeng commented:
Since there is an example added to the repo, please add the example test script here:
https://github.com/TRAIS-Lab/dattri/blob/dfdf621330e694c96e8eb085ac4ba462372db1da/.github/workflows/examples_test.yml#L22-L33

jxbb824 commented:
In my example, I used `matplotlib` for plotting. Will this cause any issues during the testing process?

TheaperDeng commented:
> In my example, I used `matplotlib` for plotting. Will this cause any issues during the testing process?

Please add any dependencies you need for testing here: https://github.com/TRAIS-Lab/dattri/blob/02aa0906388d5dec420e7366cfffd7ab77a94e69/pyproject.toml#L32

TheaperDeng commented on dattri/algorithm/base.py:
```diff
@@ -314,6 +321,20 @@ def attribute(
+                    test_batch_rep = self.generate_test_rep(
+                        ckpt_idx=checkpoint_idx,
+                        data=train_batch_data,
+                    )
```
maybe make this calculation conditional on if the `relatif_method=='l'`?

TheaperDeng reviewed (commented):
others LGTM

jxbb824 commented:
I’ve made `test_batch_rep` optional and adjusted its calculation to be conditional. I’ve also added `matplotlib` as a dependency in `pyproject.toml`, but I’m not sure if the example will pass testing since I couldn’t run it locally.
